exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/union/UnionAllRecordBatch.java

```diff
@@ -142,19 +148,26 @@ public int getRecordCount() {
 
 
   @SuppressWarnings("resource")
-  private IterOutcome doWork(RecordBatch inputBatch, boolean newSchema) throws ClassTransformationException, IOException, SchemaChangeException {
-    Preconditions.checkArgument(inputBatch.getSchema().getFieldCount() == container.getSchema().getFieldCount(),
+  private IterOutcome doWork(BatchStatusWrappper batchStatus, boolean newSchema) throws ClassTransformationException, IOException, SchemaChangeException {
+    Preconditions.checkArgument(batchStatus.batch.getSchema().getFieldCount() == container.getSchema().getFieldCount(),
         "Input batch and output batch have different field counthas!");
 
     if (newSchema) {
-      createUnionAller(inputBatch);
+      createUnionAller(batchStatus.batch);
     }
 
+    // Get number of records to include in the batch.
+    final int recordsToProcess = Math.min(batchMemoryManager.getOutputRowCount(), (batchStatus.totalRecordsToProcess - batchStatus.recordsProcessed));
```

Comment:
Minor suggestion: Add a method to ```BatchStatusWrapper``` that returns the result of this subtraction; something like ```batchStatus.remainingRecords()``` looks "cleaner"

Reply:
done.